### PR TITLE
QDEL_IN harddel prevention

### PR DESCRIPTION
**Pull request by BraveMole**
## About The Pull Request

QDEL_IN could hardel if the timer was greater than 5 minutes (time for the qdel queue)

we can use weakref to prevent this

## Changelog
:cl:
code: prevent possible QDEL_IN harddel
/:cl:

### CODE REVIEW

File: code/controllers/subsystem/garbage.dm [hunk 1]
```diff
@@ -29,7 +29,7 @@ SUBSYSTEM_DEF(garbage)
 	runlevels = RUNLEVELS_DEFAULT | RUNLEVEL_LOBBY
 	init_order = INIT_ORDER_GARBAGE
 
-	var/list/collection_timeout = list(5 MINUTES, 10 SECONDS) // deciseconds to wait before moving something up in the queue to the next level
+	var/list/collection_timeout = list(GC_QDEL_IN_CUTOFF, 10 SECONDS) // deciseconds to wait before moving something up in the queue to the next level
```
Comment (MrStonedOne) [inline code/controllers/subsystem/garbage.dm:32]: I feel like you should change the name to make more sense in this or maybe add a second define and have one use the other

File: code/controllers/subsystem/garbage.dm [hunk 2]
```diff
@@ -29,7 +29,7 @@ SUBSYSTEM_DEF(garbage)
 	runlevels = RUNLEVELS_DEFAULT | RUNLEVEL_LOBBY
 	init_order = INIT_ORDER_GARBAGE
 
-	var/list/collection_timeout = list(5 MINUTES, 10 SECONDS) // deciseconds to wait before moving something up in the queue to the next level
+	var/list/collection_timeout = list(GARBAGE_QUEUE_FIRST_LEVEL_TIMEOUT, 10 SECONDS) // deciseconds to wait before moving something up in the queue to the next level
```
Comment (LemonInTheDark) [inline code/controllers/subsystem/garbage.dm:32]: You should define both, and name them like GC_FILTER_QUEUE and GC_DEL_QUEUE or something that describes their purpose rather then just first and second

Comment (LemonInTheDark) [reply code/controllers/subsystem/garbage.dm:32]: Or maybe something else, just keep it brief

Comment (BraveMole) [inline code/controllers/subsystem/garbage.dm:32]:
```suggestion
	var/list/collection_timeout = list(GC_FILTER_QUEUE, GC_DEL_QUEUE) // deciseconds to wait before moving something up in the queue to the next level
```

File: code/__DEFINES/qdel.dm [hunk 3]
```diff
@@ -35,6 +35,9 @@
 #define GC_QUEUED_FOR_QUEUING -1
 #define GC_CURRENTLY_BEING_QDELETED -2
 
+// Defines for the time left for an item to get its reference cleaned
+#define GARBAGE_QUEUE_FIRST_LEVEL_TIMEOUT 5 MINUTES
```
Comment (BraveMole) [inline code/__DEFINES/qdel.dm:39]:
```suggestion
#define GC_FILTER_QUEUE 5 MINUTES
#define GC_DEL_QUEUE 10 SECONDS
```

File: code/__HELPERS/qdel.dm [hunk 4]
```diff
@@ -1,4 +1,4 @@
-#define QDEL_IN(item, time) addtimer(CALLBACK(GLOBAL_PROC, .proc/qdel, item), time, TIMER_STOPPABLE)
+#define QDEL_IN(item, time) addtimer(CALLBACK(GLOBAL_PROC, .proc/qdel, time > GARBAGE_QUEUE_FIRST_LEVEL_TIMEOUT ? WEAKREF(item) : item), time, TIMER_STOPPABLE)
```
Comment (BraveMole) [inline code/__HELPERS/qdel.dm:1]:
```suggestion
#define QDEL_IN(item, time) addtimer(CALLBACK(GLOBAL_PROC, .proc/qdel, time > GC_FILTER_QUEUE ? WEAKREF(item) : item), time, TIMER_STOPPABLE)
```

File: code/__DEFINES/qdel.dm (hunk 3 above)
Comment (BraveMole) [inline code/__DEFINES/qdel.dm:39]:
```suggestion
#define GC_FILTER_QUEUE 5 MINUTES
#define GC_DEL_QUEUE 10 SECONDS
```

**Review by Cyberboss — APPROVED** [review]

**Review by MrsTonedOne — APPROVED** [review]

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Review by MrsTonedOne — CHANGES_REQUESTED** [review]
fix up the qdel thing

File: code/__HELPERS/qdel.dm [hunk 5]
```diff
@@ -1,4 +1,4 @@
-#define QDEL_IN(item, time) addtimer(CALLBACK(GLOBAL_PROC, .proc/qdel, item), time, TIMER_STOPPABLE)
+#define QDEL_IN(item, time) addtimer(CALLBACK(GLOBAL_PROC, .proc/qdel, time > GC_FILTER_QUEUE ? WEAKREF(item) : item), time, TIMER_STOPPABLE)
```
[the remark this answers is not in the document]
Comment (BraveMole) [reply code/__HELPERS/qdel.dm:1]: Yes https://github.com/tgstation/tgstation/pull/56171

**Review by MrsTonedOne — APPROVED** [review]

**Pull request merged**